### ISSUE TITLE
Implement iRODS user file space in HydroShare

### ISSUE THREAD

**Issue by hyi** [hydroshare/hydroshare#1120]
This issue will implement the proposed HydroShare iRODS user zone federated with current HydroShare backend iRODS zone to better handle large file uploading, downloading, and processing. I am creating the initial task list below which is subject to further modification, adaptation, and addition based on team collective feedback and as the implementation moves forward.
- [x] Add a checkbox on user profile page so that users can indicate whether they want to create a corresponding iRODS user in HydroShare iRODS user zone. HydroShare then create or remove the corresponding iRODS users as selected by users. For this first phase of implementation, we will go with standard password user authentication in iRODS, so we will ask users to input a password for their iRODS users as well, but this password will not be saved/managed in Django at all, but instead managed by iRODS catalog for user authentication when users use external iRODS clients such as icommands or Cyberduck to log in their hydroshare user zone.
- [x] Add a "Replicate to my iRODS user zone" on resource landing page that is only enabled or visible when the checkbox above is checked that indicates a corresponding iRODS user exists for this user. This button allows users to replicate resources to their iRODS user zone. 
- [x] On resource creation page, cap file size to 1GB for Django browser file upload, and direct users to use Cyberduck or icommands (other iRODS client) to upload their larger-than-1GB files to their iRODS user zone,  Then behind the scenes, HydroShare will create resource using files uploaded in irods user zone. The resource created this way will always reference files in irods user zone without copying/moving data.  The downloaded bags for these reference iRODS resources don't content resource content files stored in iRODS user zone but only contain resource metadata xml file, resource map xml, and manifest files in the zipped bag. The resource content file location is encoded in resourcemap.xml so users can refer to it when downloading content files from their irods user zone using an external iRODS client such as icommands or cyberduck. 

In summary, we create and manage iRODS user accounts for HydroShare users from HydroShare web page, allow them to create HydroShare resources using content files stored in iRODS user zone, allow them to replicate or remove HydroShare resources from or to their irods user zone, and delegate content file upload and download to native irods clients such as icommands or Cyberduck to leverage parallel file transfer. 

As for user authentication, for this first phase implementation to get the functionality ready for use by NFIE summer institute, we just go with standard password user authentication in iRODS. For the second phase implementation, OAuth will be implemented in iRODS so that we can authenticate HydroShare users in iRODS without using native password authentication.

**Comment by hyi** [comment]
@rayi113 @mjstealey I have just updated this issue incorporating ideas in our recent discussion. Let me know if you see any problems.

**Comment by hyi** [comment]
![create-irods-account](https://cloud.githubusercontent.com/assets/476302/15194293/331481ca-1790-11e6-9430-5724b0e745e7.png)
![delete_irods_account](https://cloud.githubusercontent.com/assets/476302/15194296/35f9eb78-1790-11e6-8881-a156c99baad5.png)
@dtarb See the screen capture above for the informational language I drafted when creating or deleting an iRODS account user for a HS user. Let me know the modification you'd like to make to make it more clear to the user, and I will change the wordings accordingly.

**Comment by hyi** [comment]
@rayi113 suggested to add a message on resource landing page informing users the resource is defunct if the referenced content files have been deleted from iRODS user zone outside HydroShare resulting in broken reference links to content files, which I think is a good idea and have implemented this. When a public or discoverable resource becomes defunct, we should at least make it private and not-sharable. However, this only happens when one clicks on the resource link and gets to the landing page of this resource at which time a check is performed to see whether any referenced content file has a broken link. However, the user will still be able to retrieve such defunct resources via REST API.  @dtarb @alvacouch @pkdash Is this acceptable? I am wondering what is the best approach to handle this. 
@dtarb @alvacouch @mjstealey Also, @rayi113 and I are in agreement that allowing users to copy large files from iRODS user space to HS www zone is not a good idea since both zones share the same 50TB disk space quota, and we should not allow users to duplicate large files. So I plan to implement an option to allow users to choose to move large files from their user space (by default, no file movement with a reference link created), but no copy action will be supported.

**Comment by pkdash** [comment]
> However, the user will still be able to retrieve such defunct resources via REST API

@hyi Not sure how this is possible.

**Comment by hyi** [comment]
@pkdash @rayi113 @dtarb @alvacouch What I meant by "the user still be able to retrieve such defunct resources via REST API" is that the resource is still intact in HydroShare even some of its referenced content file links are broken (i.e., content files are deleted from user space). You can still download the bag, the "file not exist" exception only is raised when the user tries to retrieve the content files referenced from the defunct resource. Since we don't know when exactly the referenced content files are broken, there is always a chance the resource is retrieved from REST API with broken reference content file links unless we add a check. Now that I think about it more, I think we need to add such check so that REST API consumer (e.g., apps) gets notified with the resource defunct status. I am thinking then to add a flag on resource model to indicate defunct status so that we don't always have to check defunct status whenever the defunct resource is retrieved. Let me know if this still does not make sense to you or you have other ideas.

**Comment by hyi** [comment]
@pkdash @rayi113 @dtarb @alvacouch @mjstealey A supplemental info to the comment above: we only do the check when the resource has referenced files from user space and defunct flag is false whenever the resource is retrieved from REST API or resource landing page is hit.

**Comment by pkdash** [comment]
@hyi Got is as it relates to REST API being able to get the bag for a defunct resource. In the case of such a resource I assume we won't be displaying content file links (for broken links) for download or delete on resource landing page. I am wondering how the user be able to update the content files for such a resource then.

For sure it is going to add complexity to the workflow of managing resources in Hydroshare. I would prefer to copy the file so that we don't have to do all these things. Can't we put a quota on user space?

**Comment by hyi** [comment]
@pkdash Right, on resource landing page, although we should the broken link, we add informational message informing the user this link is broken and this defunct resource should be deleted, and download button is disabled with informational tooltip to inform users the reason. Other related functions such as deleting resource file will also be disabled or not available for the defunct resources. I am thinking perhaps we should make such defunct resource non-editable except for deleting it. Since access control does not have this logic built in, it will add complexity to the workflow of managing resources. Another option is to add a check in the cron-type celery task (we already have one for resource publication check) for such defunct resources and delete defunct resources automatically in the nightly-run celery task and email the owners to inform the defunct status of their resources which have been deleted. I like this option better which is simplest to implement and sounds reasonable to me. What do you all think about this? @rayi113 @dtarb @alvacouch @mjstealey @pkdash

**Comment by rayi113** [comment]
@pkdash The reason that this is not a quota issue is that one of the use cases that Dave @dtarb and I discussed at the NSF SI2 PI workshop was enabling users to bring their own (iRODS) disk space to HydroShare.  As a policy, we won't be federating these or adding these directly to hydroshareZone, so what we decided made the most sense was for users to federate or add these as iRODS resource servers to the HydroShare user zone.  This use case also necessitates leaving files in place on these federated or additional resource servers (because it is the external disk space that must be used) thus necessitating the model and details that Hong @hyi puts forth.  @alvacouch @mjstealey

**Comment by pkdash** [comment]
@rayi113 So why is then copying the file a problem if disk space is not an issue?

**Comment by pkdash** [comment]
@rayi113 If the requirement is to create reference to files outside the hydroshareZone, then copying is not an option.

**Comment by rayi113** [comment]
@pkdash Disk space is actually the issue, but not from a quota standpoint, rather from the use case that necessitates a HydroShare user bringing their own disk space to HydroShare.  Whether we limit that user to 1TB or 10TB via a quota on HydroShare core disk space does not address this use case since the story here is how to technologically implement a HydroShare user bringing their own disk space to HydroShare.  If that user brings 100TB of their own disk to HydroShare (actually the National Water Model data is almost this use case), and we only have 50TB in HydroShare, copying is not a viable paradigm because we must enable publishing from their external disk space and not ours but also conceding there will not be enough core HydroShare disk space to support this use case, and this is what we want to avoid and address.  @dtarb @hyi @mjstealey @alvacouch

**Comment by pkdash** [comment]
@hyi @rayi113 Where the bags for these resources would exist? Not in hydroshareZone?

**Comment by hyi** [comment]
@pkdash The bags for these resources exist in hydroshareZone, only content files are referenced from user zone. I intended to demo the bag content and resourcemap.xml for such resources with referenced content files in user zone, but did not get to it due to time limit. I plan to continue the demo at our next team call meeting. Also, @rayi113 brought up a good case above for 100TB of their own data disk with NWM data, which makes me think whether we really want to provide option to allow users to even move their data to hydroshareZone if the data is really big. I will not implement this option until we make a decision on whether we want to provide this option or not given its purpose for large file handling. I will go with nightly celery task check for defunct resources and delete them automatically approach, but do want to show defunct status if users hit those resource landing page or retrieve such resources via REST API, unless I hear objections. @dtarb @alvacouch @rayi113 let me know if you oppose this approach. Also I will not implement "move file" option as I originally proposed due to the concern discussed above. Let me know what you think.

**Comment by pkdash** [comment]
@hyi Does it mean then the bag won't have the content files for these kind of resources? Similar to collection resource bag.

**Comment by hyi** [comment]
@pkdash Right, the bag does not have the content files for these reference resources, but will have full reference path encoded as URI in resourcemap.xml, which is resolvable to download the referenced files if the files are smaller than 1GB, otherwise, if the files are larger than 1GB which cannot be handled well by http download any way, users will have to download the file via iRODS directly from their user space, but the URI in resourcemap.xml is resolvable.

**Comment by pkdash** [comment]
Sounds good

**Comment by dtarb** [comment]
@hyi, I have a concern that I think we need strict separation between files in bags that are part of resources and files that are in a place where a user has uploaded them and can work on and delete them.  The way this seems to be being built, a user can add a file to iRODS, create a resource, then delete the file and the resource is broken.  We do not want to do this.

I developed the model for iRODS enhancements described in the Diagram https://docs.google.com/drawings/d/1e0YuYUdq8AfIIxELen9MdN3_qkcnrILy-E-LMbl30JA/edit a while back.  I realize that subsequent discussions may have superceded this diagram, but I think that the key ideas still apply and we need an updated diagram that conveys the complete picture.

One of the changes following the above design is that the distributed iRODS servers need to not be in the same iRODS zone as that would require the use of HydroShare admin credentials by the remote site, a protocol we do not want to do for security reasons.  However I think that almost everything else can apply.  Remote iRODS resource servers that are federated with HydroShare should have a HydroShare proxy user, just like the HydroShare zone has a HydroShare proxy user.  This can be a requirement for federating with HydroShare.  Then this HydroShare proxy user is the holder of resource files (bags).   When a user creates a hydroshare resource on that server the default mode should be to move the file from its local user place in iRODS to the proxy user HS resource files.  The file then disappears from user iRODS control and is under HS proxy iRODS control so that integrity of resources and bags can be maintained.  The file is still resident on the federated server.

The important distinction here is that I think there is a greater level of stability of a federated iRODS resource server, than just a user with a file in their folder, so it is OK to support and publish resources from a federated iRODS system, but not OK just from user space.

**Comment by hyi** [comment]
@dtarb Agreed that moving files from irods user space to hydroshare zone makes more sense. This way, we treat irods user space as a temporary staging area and once the HS resource is created successfully, the staging big files in the user space is moved to HydroShare zone. Since it is staging area, I think users should be able to accept the fact that their files are gone from their user space after the files are contained in hydroshare resources. This way we don't have to deal with broken resource issues, and we will not have duplicate big file issue either. The only issue I see with this approach is that the user may want to do work with files in their user space even after creating a hydroshare resource, e.g., run models directly using those files in their user space. However, I did add a button on resource landing page that allows users to replicate the resource bag to their user space for downloading (http download is capped to 1GB file size) or for doing works with resource bag in their irods user space, so that may not be an issue after all. It appears to me that this "moving files from irods user zone to hydroshare zone" makes more sense and is what we should go with. @dtarb @rayi113 @alvacouch @pkdash @mjstealey thoughts?

**Comment by hyi** [comment]
@dtarb @rayi113 @alvacouch @pkdash @mjstealey I come up with a hopefully better approach from user perspective for this issue. Specifically, we still go with current implementation to create reference links to content files in irods user zone/space by default when creating a HS resource, however, we will make the content files in iRODS user zone read-only after the referenced HS resource is creating, which will disable content files from being deleted outside HydroShare, as such we will not have this broken resource link issue. When the HS resource is deleted, the content files it references in iRODS user zone will not be deleted, but instead their original permissions are restored so that users can have full permission to their files in their user space again once they are not referenced in HS any more. I will also implement a MOVE option that allows users to move files from their user space to HydroShare proxy space if they choose to, but by default the referenced resource will be created without data movement. This seems to be the most efficient approach to manage large files without unnecessary big data movement unless the user chooses to. Also, we will not provide copy options to users since we should not encourage duplication of big data. I plan to go with this approach unless I hear concerns or objections. Let me know your thoughts.

**Comment by dtarb** [comment]
I think these suggestions seem generally good for within the HydroShare iRODS zone.

You mentioned several comments above the use case of a user (really organization) that brings their own disk.  This would be an organization like NCSA or National Water Model that is federating with HydroShare.  Will this approach work within a federated iRODS server that is under their control.  I suspect it may be difficult to have rules set up that manage the permissions in a user space on a federated server, so suggest then that the federated server option be implemented with a HydroShare proxy account on that server and moving (without copying) the resource files into the HydroShare space owned by that account.

**Comment by hyi** [comment]
@dtarb Yes, the suggestion above only works well and makes sense for hydroshare user zone/space within our control which is what I am currently implementing. Agreed it will not work well for the use case of an organization that brings their own disk to federate with HydroShare. I'll defer this use case to @rayi113 on how to best go about it on iRODS federation level, and we will have to handle it a bit differently from the use case of hydroshare user zone/space.

**Comment by hyi** [comment]
Closing this issue since irods user file space has been implemented and deployed. Remaining issues to be worked on are created as #1295 and #1292
